sql/pipelines/src/test/scala/org/apache/spark/sql/pipelines/utils/PipelineTest.scala

```diff
@@ -0,0 +1,461 @@
+      log.info(s"Spark plan:\n${df.queryExecution.executedPlan}")
+    }
+    checkPlan.foreach(_.apply(df.queryExecution.executedPlan))
+  }
+
+  /**
+   * Runs the plan and makes sure the answer matches the expected result.
+   *
+   * @param df the [[DataFrame]] to be executed
+   * @param expectedAnswer the expected result in a [[Seq]] of [[Row]]s.
+   */
+  protected def checkAnswer(df: => DataFrame, expectedAnswer: Seq[Row]): Unit = {
+    checkAnswerAndPlan(df, expectedAnswer, None)
+  }
+
+  protected def checkAnswer(df: => DataFrame, expectedAnswer: Row): Unit = {
+    checkAnswer(df, Seq(expectedAnswer))
+  }
+
+  case class ValidationArgs(
+      ignoreFieldOrder: Boolean = false,
+      ignoreFieldCase: Boolean = false
+  )
+
+  /**
+   * Evaluates a dataset to make sure that the result of calling collect matches the given
+   * expected answer.
+   */
+  protected def checkDataset[T](ds: => Dataset[T], expectedAnswer: T*): Unit = {
+    val result = getResult(ds)
+
+    if (!QueryTest.compare(result.toSeq, expectedAnswer)) {
+      fail(s"""
+              |Decoded objects do not match expected objects:
+              |expected: $expectedAnswer
+              |actual:   ${result.toSeq}
+         """.stripMargin)
+    }
+  }
+
+  /**
+   * Evaluates a dataset to make sure that the result of calling collect matches the given
+   * expected answer, after sort.
+   */
+  protected def checkDatasetUnorderly[T: Ordering](result: Array[T], expectedAnswer: T*): Unit = {
+    if (!QueryTest.compare(result.toSeq.sorted, expectedAnswer.sorted)) {
+      fail(s"""
+              |Decoded objects do not match expected objects:
+              |expected: $expectedAnswer
+              |actual:   ${result.toSeq}
+         """.stripMargin)
+    }
+  }
+
+  protected def checkDatasetUnorderly[T: Ordering](ds: => Dataset[T], expectedAnswer: T*): Unit = {
+    val result = getResult(ds)
+    if (!QueryTest.compare(result.toSeq.sorted, expectedAnswer.sorted)) {
+      fail(s"""
+              |Decoded objects do not match expected objects:
+              |expected: $expectedAnswer
+              |actual:   ${result.toSeq}
+         """.stripMargin)
+    }
+  }
+
+  private def getResult[T](ds: => Dataset[T]): Array[T] = {
+    ds
+
+    try ds.collect()
+    catch {
+      case NonFatal(e) =>
+        fail(
+          s"""
+             |Exception collecting dataset as objects
+             |${ds.queryExecution}
+           """.stripMargin,
+          e
+        )
+    }
+  }
+
+  /** Holds a parsed version along with the original json of a test. */
+  private case class TestSequence(json: Seq[String], rows: Seq[Row]) {
+    require(json.size == rows.size)
+  }
+
+  /**
+   * Helper method to verify unresolved column error message. We expect three elements to be present
+   * in the message: error class, unresolved column name, list of suggested columns. There are three
+   * significant differences between different versions of DBR:
```

Comment:
@aakash-db Here are some descriptions related to `DBR`, but since the current project is Apache Spark, could you revise these descriptions accordingly?

Reply:
Fixed here: https://github.com/apache/spark/pull/51180/files